File: data/push_pack.py
import json, os, sys, urllib.request

BASE_URL = os.environ['BASE_URL'].rstrip('/')
ADMIN_USER = os.environ['ADMIN_USER']
ADMIN_PASS = os.environ['ADMIN_PASS']
pack_file = sys.argv[1]

def call(path, method='GET', body=None, token=None):
    url = BASE_URL + path
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(url, data=data, method=method)
    req.add_header('Content-Type', 'application/json')
    if token:
        req.add_header('Authorization', f'Bearer {token}')
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read().decode())

pack = json.load(open(pack_file, encoding='utf-8'))

login = call('/api/auth/login', 'POST', {'username': ADMIN_USER, 'password': ADMIN_PASS})
token = login['token']
if not login['user']['is_admin']:
    raise SystemExit('This account is not an admin.')

existing = call('/api/admin/decks', token=token)['decks']
existing_by_name = {d['name']: d['id'] for d in existing}

total = 0
for deck in pack['decks']:
    name = deck['name']
    if name in existing_by_name:
        deck_id = existing_by_name[name]
        print(f'Using existing deck "{name}" (id {deck_id})')
    else:
        res = call('/api/admin/decks', 'POST', {'name': name, 'description': deck.get('description', '')}, token)
        deck_id = res['id']
        print(f'Created deck "{name}" (id {deck_id})')

    cards = [{'welsh': w, 'english': e} for w, e in deck['cards']]
    res = call('/api/admin/import', 'POST', {'deck_id': deck_id, 'cards': cards}, token)
    print(f'  -> imported {res["imported"]} cards')
    total += res['imported']

print(f'\nDone. Total cards imported: {total}')

File: data/push_pack.js
// Pushes a word pack (JSON: { decks: [{ name, description, cards: [[welsh, english], ...] }] })
// to a live deployed instance via the admin API.
//
// Usage:
//   BASE_URL=https://your-app.up.railway.app ADMIN_USER=... ADMIN_PASS=... \
//   node data/push_pack.js data/pack_basics_500.json

const fs = require('fs');
const path = require('path');

const BASE_URL = process.env.BASE_URL;
const ADMIN_USER = process.env.ADMIN_USER;
const ADMIN_PASS = process.env.ADMIN_PASS;
const packFile = process.argv[2];

if (!BASE_URL || !ADMIN_USER || !ADMIN_PASS || !packFile) {
  console.error('Usage: BASE_URL=... ADMIN_USER=... ADMIN_PASS=... node data/push_pack.js <pack.json>');
  process.exit(1);
}

const pack = JSON.parse(fs.readFileSync(path.resolve(packFile), 'utf8'));

async function main() {
  const loginRes = await fetch(`${BASE_URL}/api/auth/login`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ username: ADMIN_USER, password: ADMIN_PASS })
  });
  const loginData = await loginRes.json();
  if (!loginRes.ok) throw new Error('Login failed: ' + loginData.error);
  const token = loginData.token;
  if (!loginData.user.is_admin) throw new Error('This account is not an admin.');

  const headers = { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` };

  const existingDecksRes = await fetch(`${BASE_URL}/api/admin/decks`, { headers });
  const { decks: existingDecks } = await existingDecksRes.json();

  let totalImported = 0;

  for (const deck of pack.decks) {
    let deckEntry = existingDecks.find(d => d.name === deck.name);
    let deckId;
    if (deckEntry) {
      deckId = deckEntry.id;
      console.log(`Using existing deck "${deck.name}" (id ${deckId})`);
    } else {
      const res = await fetch(`${BASE_URL}/api/admin/decks`, {
        method: 'POST',
        headers,
        body: JSON.stringify({ name: deck.name, description: deck.description || '' })
      });
      const data = await res.json();
      if (!res.ok) throw new Error(`Failed to create deck "${deck.name}": ${data.error}`);
      deckId = data.id;
      console.log(`Created deck "${deck.name}" (id ${deckId})`);
    }

    const cards = deck.cards.map(([welsh, english]) => ({ welsh, english }));
    const importRes = await fetch(`${BASE_URL}/api/admin/import`, {
      method: 'POST',
      headers,
      body: JSON.stringify({ deck_id: deckId, cards })
    });
    const importData = await importRes.json();
    if (!importRes.ok) throw new Error(`Import failed for "${deck.name}": ${importData.error}`);
    console.log(`  -> imported ${importData.imported} cards`);
    totalImported += importData.imported;
  }

  console.log(`\nDone. Total cards imported: ${totalImported}`);
}

main().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});
